src/Mvc/Mvc.Core/src/Authorization/AuthorizeFilter.cs

```diff
@@ -154,6 +153,27 @@ private async Task<AuthorizationPolicy> GetEffectivePolicyAsync(AuthorizationFil
                 }
             }
 
+            var endpoint = context.HttpContext.GetEndpoint();
+            if (endpoint != null)
+            {
+                // When doing endpoint routing, MVC does not create filters for any authorization specific metadata i.e [Authorize] does not
+                // get translated into AuthorizeFilter. Consequently, there are some rough edges when an application uses a mix of AuthorizeFilter
+                // explicilty configured by the user (e.g. global auth filter), and uses endpoint metadata.
+                // To keep the behavior of AuthFilter identical to pre-endpoint routing, we will gather auth data from endpoint metadata
+                // and produce a policy using this. This would mean we would have effectively run some auth twice, but it maintains compat.
+                var policyProvider = PolicyProvider ?? context.HttpContext.RequestServices.GetRequiredService<IAuthorizationPolicyProvider>();
+                var endpointAuthorizeData = endpoint.Metadata.GetOrderedMetadata<IAuthorizeData>() ?? Array.Empty<IAuthorizeData>();
+
+                var endpointPolicy = await AuthorizationPolicy.CombineAsync(policyProvider, endpointAuthorizeData);
+                if (endpointPolicy != null)
+                {
+                    builder.Combine(endpointPolicy);
+                }
+
+                // We cannot cache the policy since it varies by endpoint metadata.
+                canCache = false;
```

Comment:
😢

Reply:
meh.

Reply:
On the bright side, there might have been some subtle bugs with policy caching + combining of policies that were in 2.2 (but appear to have been accidentally fixed in 3.0) i.e. https://github.com/aspnet/AspNetCore/issues/7687

So its slower, but safer